File: app/src/main/java/com/saud/taskstrip/sync/SyncBoardDocument.kt
package com.saud.taskstrip.sync

import org.json.JSONArray
import org.json.JSONObject

/** One strip, as it travels. Fields are the ones both apps can hold; see TaskEntity and
 * TaskItem.swift, which this sits between. Attachments travel as content hashes — the bytes
 * themselves are a separate errand, so a record stays small and a photo is uploaded once however
 * many strips point at it. */
data class SyncTaskRecord(
    val id: String,
    val updatedAt: Long = 0L,
    val isDeleted: Boolean = false,
    val title: String = "",
    val notes: String = "",
    val notesRtl: Boolean = false,
    val priority: String = "NORMAL",
    val dueAt: Long? = null,
    val orderIndex: Int = 0,
    val isDone: Boolean = false,
    val isArchived: Boolean = false,
    val progress: Int = 0,
    val completedAt: Long? = null,
    /** The blocker's shared id, not its local key — the local one means nothing on the other
     * device, which is the whole reason this document exists. */
    val blockedBySyncId: String? = null,
    val waitingOnName: String = "",
    val waitingOnSince: Long? = null,
    val waitingOnFollowUpDays: Int? = null,
    val reminderMinutesBefore: Int? = null,
    val repeatIntervalDays: Int? = null,
    val tags: List<String> = emptyList(),
    val links: List<SyncLink> = emptyList(),
    val actionLog: List<SyncLogEntry> = emptyList(),
    val contacts: List<SyncContact> = emptyList(),
    val attachments: List<SyncAttachment> = emptyList(),
    val createdAt: Long = 0L
)

data class SyncReminderRecord(
    val id: String,
    val updatedAt: Long = 0L,
    val isDeleted: Boolean = false,
    val text: String = "",
    val details: String = "",
    val triggerAt: Long = 0L,
    val leadMinutesBefore: Int? = null,
    val repeatAmount: Int? = null,
    val repeatUnit: String? = null,
    val tag: String = "",
    val tagEmoji: String = "",
    val isDone: Boolean = false,
    val createdAt: Long = 0L
)

data class SyncLink(val url: String = "", val label: String = "")
data class SyncLogEntry(val text: String = "", val timestamp: Long = 0L)
data class SyncContact(val name: String = "", val email: String = "", val phone: String = "")

/** A file a strip carries, named by what's in it.
 *
 * `hash` is the SHA-256 of the bytes, which makes the name a fact about the content rather than a
 * choice either device made. Two devices that add the same photo produce the same name and upload
 * it once; a file that changes produces a different name, so a binary never has to be merged and
 * "which copy wins" never comes up. `kind` says which of a strip's lists it belongs in. */
data class SyncAttachment(val hash: String = "", val name: String = "", val kind: String = "")

/** The board both apps read and write: one JSON file beside sync_notes.json in the folder they
 * already share.
 *
 * This file is a contract with SyncBoardDocument.swift, and every rule below has to hold
 * identically there — the two devices merge the same document independently and have to reach the
 * same answer without talking to each other. It is deliberately the same design as
 * SyncNoteDocument, which has been carrying this for a while: newest wins, a delete outranks a
 * stale edit, and the tie-breaks are there so two devices holding genuinely different text stamped
 * at the same millisecond still pick the *same* side and stop disagreeing. */
object SyncBoardDocument {
    const val FILE_NAME = "sync_board.json"
    const val MIME_TYPE = "application/json"
    const val VERSION = 1

    // ---- Reading and writing ----

    fun toJson(tasks: List<SyncTaskRecord>, reminders: List<SyncReminderRecord>): String {
        val root = JSONObject()
        root.put("version", VERSION)
        root.put("tasks", JSONArray().apply { sortedTasks(tasks).forEach { put(taskToJson(it)) } })
        root.put("reminders", JSONArray().apply { sortedReminders(reminders).forEach { put(reminderToJson(it)) } })
        return root.toString(2)
    }

    /** Anything unreadable is nothing rather than an error. A corrupt or half-written document
     * must not be able to empty this device — the merge treats "no remote" as "nothing to add",
     * and the next upload writes a good document over it. */
    fun tasksFromJson(json: String): List<SyncTaskRecord> = runCatching {
        val array = JSONObject(json).optJSONArray("tasks") ?: return emptyList()
        (0 until array.length()).mapNotNull { i ->
            val obj = array.optJSONObject(i) ?: return@mapNotNull null
            val id = obj.optString("id")
            if (id.isNullOrEmpty()) return@mapNotNull null
            SyncTaskRecord(
                id = id,
                updatedAt = obj.optLong("updatedAt", 0L),
                isDeleted = obj.optBoolean("deleted", false),
                title = obj.optString("title", ""),
                notes = obj.optString("notes", ""),
                notesRtl = obj.optBoolean("notesRtl", false),
                priority = obj.optString("priority", "NORMAL").ifBlank { "NORMAL" },
                dueAt = obj.optLongOrNull("dueAt"),
                orderIndex = obj.optInt("orderIndex", 0),
                isDone = obj.optBoolean("done", false),
                isArchived = obj.optBoolean("archived", false),
                progress = obj.optInt("progress", 0),
                completedAt = obj.optLongOrNull("completedAt"),
                blockedBySyncId = obj.optStringOrNull("blockedBy"),
                waitingOnName = obj.optString("waitingOnName", ""),
                waitingOnSince = obj.optLongOrNull("waitingOnSince"),
                waitingOnFollowUpDays = obj.optIntOrNull("waitingOnFollowUpDays"),
                reminderMinutesBefore = obj.optIntOrNull("reminderMinutesBefore"),
                repeatIntervalDays = obj.optIntOrNull("repeatIntervalDays"),
                tags = obj.stringList("tags"),
                links = obj.objectList("links") { SyncLink(it.optString("url", ""), it.optString("label", "")) },
                actionLog = obj.objectList("actionLog") { SyncLogEntry(it.optString("text", ""), it.optLong("timestamp", 0L)) },
                contacts = obj.objectList("contacts") {
                    SyncContact(it.optString("name", ""), it.optString("email", ""), it.optString("phone", ""))
                },
                attachments = obj.objectList("attachments") {
                    SyncAttachment(it.optString("hash", ""), it.optString("name", ""), it.optString("kind", ""))
                },
                createdAt = obj.optLong("createdAt", 0L)
            )
        }
    }.getOrDefault(emptyList())

    fun remindersFromJson(json: String): List<SyncReminderRecord> = runCatching {
        val array = JSONObject(json).optJSONArray("reminders") ?: return emptyList()
        (0 until array.length()).mapNotNull { i ->
            val obj = array.optJSONObject(i) ?: return@mapNotNull null
            val id = obj.optString("id")
            if (id.isNullOrEmpty()) return@mapNotNull null
            SyncReminderRecord(
                id = id,
                updatedAt = obj.optLong("updatedAt", 0L),
                isDeleted = obj.optBoolean("deleted", false),
                text = obj.optString("text", ""),
                details = obj.optString("details", ""),
                triggerAt = obj.optLong("triggerAt", 0L),
                leadMinutesBefore = obj.optIntOrNull("leadMinutesBefore"),
                repeatAmount = obj.optIntOrNull("repeatAmount"),
                repeatUnit = obj.optStringOrNull("repeatUnit"),
                tag = obj.optString("tag", ""),
                tagEmoji = obj.optString("tagEmoji", ""),
                isDone = obj.optBoolean("done", false),
                createdAt = obj.optLong("createdAt", 0L)
            )
        }
    }.getOrDefault(emptyList())

    // ---- Merging ----

    fun mergeTasks(local: List<SyncTaskRecord>, remote: List<SyncTaskRecord>): List<SyncTaskRecord> {
        val byId = LinkedHashMap<String, SyncTaskRecord>()
        (local + remote).forEach { record ->
            val existing = byId[record.id]
            byId[record.id] = if (existing == null) record else winner(existing, record)
        }
        return sortedTasks(byId.values.toList())
    }

    fun mergeReminders(local: List<SyncReminderRecord>, remote: List<SyncReminderRecord>): List<SyncReminderRecord> {
        val byId = LinkedHashMap<String, SyncReminderRecord>()
        (local + remote).forEach { record ->
            val existing = byId[record.id]
            byId[record.id] = if (existing == null) record else winner(existing, record)
        }
        return sortedReminders(byId.values.toList())
    }

    /** Newer wins. Then a delete wins over a live row, because a delete is a decision and a stale
     * edit is not. Then the greater title, then the greater notes, purely so two devices holding
     * genuinely different rows stamped at the same millisecond still pick the *same* side.
     *
     * Two edits in the same millisecond that differ only past those fields keep whichever side the
     * caller had first. That is a real limit and an unreachable one in practice: it needs both
     * devices to write the same strip within the same millisecond and to agree on its title and
     * notes while disagreeing elsewhere. */
    fun winner(a: SyncTaskRecord, b: SyncTaskRecord): SyncTaskRecord {
        if (a.updatedAt != b.updatedAt) return if (a.updatedAt > b.updatedAt) a else b
        if (a.isDeleted != b.isDeleted) return if (a.isDeleted) a else b
        if (a.title != b.title) return if (isGreater(a.title, b.title)) a else b
        if (a.notes != b.notes) return if (isGreater(a.notes, b.notes)) a else b
        return a
    }

    fun winner(a: SyncReminderRecord, b: SyncReminderRecord): SyncReminderRecord {
        if (a.updatedAt != b.updatedAt) return if (a.updatedAt > b.updatedAt) a else b
        if (a.isDeleted != b.isDeleted) return if (a.isDeleted) a else b
        if (a.text != b.text) return if (isGreater(a.text, b.text)) a else b
        if (a.details != b.details) return if (isGreater(a.details, b.details)) a else b
        return a
    }

    /** Compares UTF-8 bytes, not strings — the same rule and the same reason as
     * SyncNoteDocument.isGreater: Kotlin orders UTF-16 code units and Swift orders grapheme
     * clusters, they agree for ASCII and need not for Arabic, and a tie-break the two platforms
     * disagree about is worse than no tie-break at all. */
    fun isGreater(a: String, b: String): Boolean {
        val x = a.toByteArray(Charsets.UTF_8)
        val y = b.toByteArray(Charsets.UTF_8)
        for (i in 0 until minOf(x.size, y.size)) {
            val left = x[i].toInt() and 0xFF
            val right = y[i].toInt() and 0xFF
            if (left != right) return left > right
        }
        return x.size > y.size
    }

    /** Ordered by id alone, so the file's bytes don't churn between syncs that changed nothing.
     * A strip's own orderIndex is what the board reads; this is only the document's order. */
    fun sortedTasks(tasks: List<SyncTaskRecord>): List<SyncTaskRecord> = tasks.sortedBy { it.id }

    fun sortedReminders(reminders: List<SyncReminderRecord>): List<SyncReminderRecord> =
        reminders.sortedBy { it.id }

    /** What a person should see: tombstones are bookkeeping, not rows. */
    fun visibleTasks(tasks: List<SyncTaskRecord>): List<SyncTaskRecord> = tasks.filter { !it.isDeleted }

    fun visibleReminders(reminders: List<SyncReminderRecord>): List<SyncReminderRecord> =
        reminders.filter { !it.isDeleted }

    /** Every file the board still points at. What isn't in here is an orphan and can go. */
    fun referencedHashes(tasks: List<SyncTaskRecord>): Set<String> =
        tasks.filter { !it.isDeleted }.flatMap { task -> task.attachments.map { it.hash } }
            .filter { it.isNotEmpty() }
            .toSet()

    // ---- JSON plumbing ----

    private fun taskToJson(task: SyncTaskRecord) = JSONObject().apply {
        put("id", task.id)
        put("updatedAt", task.updatedAt)
        put("deleted", task.isDeleted)
        put("title", task.title)
        put("notes", task.notes)
        put("notesRtl", task.notesRtl)
        put("priority", task.priority)
        putOrNull("dueAt", task.dueAt)
        put("orderIndex", task.orderIndex)
        put("done", task.isDone)
        put("archived", task.isArchived)
        put("progress", task.progress)
        putOrNull("completedAt", task.completedAt)
        putOrNull("blockedBy", task.blockedBySyncId)
        put("waitingOnName", task.waitingOnName)
        putOrNull("waitingOnSince", task.waitingOnSince)
        putOrNull("waitingOnFollowUpDays", task.waitingOnFollowUpDays)
        putOrNull("reminderMinutesBefore", task.reminderMinutesBefore)
        putOrNull("repeatIntervalDays", task.repeatIntervalDays)
        put("tags", JSONArray(task.tags))
        put("links", JSONArray().apply {
            task.links.forEach { put(JSONObject().put("url", it.url).put("label", it.label)) }
        })
        put("actionLog", JSONArray().apply {
            task.actionLog.forEach { put(JSONObject().put("text", it.text).put("timestamp", it.timestamp)) }
        })
        put("contacts", JSONArray().apply {
            task.contacts.forEach {
                put(JSONObject().put("name", it.name).put("email", it.email).put("phone", it.phone))
            }
        })
        put("attachments", JSONArray().apply {
            task.attachments.forEach {
                put(JSONObject().put("hash", it.hash).put("name", it.name).put("kind", it.kind))
            }
        })
        put("createdAt", task.createdAt)
    }

    private fun reminderToJson(reminder: SyncReminderRecord) = JSONObject().apply {
        put("id", reminder.id)
        put("updatedAt", reminder.updatedAt)
        put("deleted", reminder.isDeleted)
        put("text", reminder.text)
        put("details", reminder.details)
        put("triggerAt", reminder.triggerAt)
        putOrNull("leadMinutesBefore", reminder.leadMinutesBefore)
        putOrNull("repeatAmount", reminder.repeatAmount)
        putOrNull("repeatUnit", reminder.repeatUnit)
        put("tag", reminder.tag)
        put("tagEmoji", reminder.tagEmoji)
        put("done", reminder.isDone)
        put("createdAt", reminder.createdAt)
    }
}

// A null has to be written as JSON null, not left out and not turned into the string "null" —
// JSONObject.put(String, Object?) removes the key, and a reader elsewhere would then see a
// default where the writer meant "nothing".
private fun JSONObject.putOrNull(key: String, value: Any?) {
    if (value == null) put(key, JSONObject.NULL) else put(key, value)
}

private fun JSONObject.optLongOrNull(key: String): Long? =
    if (isNull(key)) null else optLong(key)

private fun JSONObject.optIntOrNull(key: String): Int? =
    if (isNull(key)) null else optInt(key)

private fun JSONObject.optStringOrNull(key: String): String? =
    if (isNull(key)) null else optString(key).ifEmpty { null }

private fun JSONObject.stringList(key: String): List<String> {
    val array = optJSONArray(key) ?: return emptyList()
    return (0 until array.length()).map { array.optString(it, "") }.filter { it.isNotEmpty() }
}

private fun <T> JSONObject.objectList(key: String, build: (JSONObject) -> T): List<T> {
    val array = optJSONArray(key) ?: return emptyList()
    return (0 until array.length()).mapNotNull { array.optJSONObject(it)?.let(build) }
}

File: macos/TaskStrips/Sources/TaskStrips/Sync/SyncBoardDocument.swift
import Foundation

/// One strip, as it travels. Mirrors SyncTaskRecord in SyncBoardDocument.kt field for field.
///
/// Attachments travel as content hashes; the bytes are a separate errand, so a record stays small
/// and a photo is uploaded once however many strips point at it.
struct SyncTaskRecord: Equatable, Identifiable {
    var id: String
    var updatedAt: Int64 = 0
    var isDeleted = false
    var title = ""
    var notes = ""
    var notesRtl = false
    var priority = "NORMAL"
    var dueAt: Int64?
    var orderIndex = 0
    var isDone = false
    var isArchived = false
    var progress = 0
    var completedAt: Int64?
    /// The blocker's shared id, not its local key — a local key means nothing on the other device,
    /// which is the whole reason this document exists.
    var blockedBySyncID: String?
    var waitingOnName = ""
    var waitingOnSince: Int64?
    var waitingOnFollowUpDays: Int?
    var reminderMinutesBefore: Int?
    var repeatIntervalDays: Int?
    var tags: [String] = []
    var links: [SyncLink] = []
    var actionLog: [SyncLogEntry] = []
    var contacts: [SyncContact] = []
    var attachments: [SyncAttachment] = []
    var createdAt: Int64 = 0
}

struct SyncReminderRecord: Equatable, Identifiable {
    var id: String
    var updatedAt: Int64 = 0
    var isDeleted = false
    var text = ""
    var details = ""
    var triggerAt: Int64 = 0
    var leadMinutesBefore: Int?
    var repeatAmount: Int?
    var repeatUnit: String?
    var tag = ""
    var tagEmoji = ""
    var isDone = false
    var createdAt: Int64 = 0
}

struct SyncLink: Equatable { var url = ""; var label = "" }
struct SyncLogEntry: Equatable { var text = ""; var timestamp: Int64 = 0 }
struct SyncContact: Equatable { var name = ""; var email = ""; var phone = "" }

/// A file a strip carries, named by what's in it.
///
/// `hash` is the SHA-256 of the bytes, which makes the name a fact about the content rather than a
/// choice either device made. Two devices that add the same photo produce the same name and upload
/// it once; a file that changes produces a different name, so a binary never has to be merged and
/// "which copy wins" never comes up.
struct SyncAttachment: Equatable { var hash = ""; var name = ""; var kind = "" }

/// The board both apps read and write: one JSON file beside sync_notes.json in the folder they
/// already share.
///
/// A contract with SyncBoardDocument.kt — the two devices merge the same document independently
/// and have to reach the same answer without talking to each other. Deliberately the same design
/// as SyncNoteDocument, which has been carrying this for a while.
enum SyncBoardDocument {
    static let fileName = "sync_board.json"
    static let mimeType = "application/json"
    static let version = 1

    // MARK: - Reading and writing

    static func data(tasks: [SyncTaskRecord], reminders: [SyncReminderRecord]) throws -> Data {
        let root: [String: Any] = [
            "version": version,
            "tasks": sorted(tasks).map(json(for:)),
            "reminders": sorted(reminders).map(json(for:)),
        ]
        return try JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted, .sortedKeys])
    }

    /// Anything unreadable is nothing rather than an error. A corrupt or half-written document
    /// must not be able to empty this machine — the merge treats "no remote" as "nothing to add".
    static func tasks(from data: Data) -> [SyncTaskRecord] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let array = root["tasks"] as? [[String: Any]] else { return [] }
        return array.compactMap { object in
            guard let id = object["id"] as? String, !id.isEmpty else { return nil }
            return SyncTaskRecord(
                id: id,
                updatedAt: int64(object["updatedAt"]) ?? 0,
                isDeleted: object["deleted"] as? Bool ?? false,
                title: object["title"] as? String ?? "",
                notes: object["notes"] as? String ?? "",
                notesRtl: object["notesRtl"] as? Bool ?? false,
                priority: (object["priority"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "NORMAL",
                dueAt: int64(object["dueAt"]),
                orderIndex: int(object["orderIndex"]) ?? 0,
                isDone: object["done"] as? Bool ?? false,
                isArchived: object["archived"] as? Bool ?? false,
                progress: int(object["progress"]) ?? 0,
                completedAt: int64(object["completedAt"]),
                blockedBySyncID: string(object["blockedBy"]),
                waitingOnName: object["waitingOnName"] as? String ?? "",
                waitingOnSince: int64(object["waitingOnSince"]),
                waitingOnFollowUpDays: int(object["waitingOnFollowUpDays"]),
                reminderMinutesBefore: int(object["reminderMinutesBefore"]),
                repeatIntervalDays: int(object["repeatIntervalDays"]),
                tags: (object["tags"] as? [String] ?? []).filter { !$0.isEmpty },
                links: (object["links"] as? [[String: Any]] ?? []).map {
                    SyncLink(url: $0["url"] as? String ?? "", label: $0["label"] as? String ?? "")
                },
                actionLog: (object["actionLog"] as? [[String: Any]] ?? []).map {
                    SyncLogEntry(text: $0["text"] as? String ?? "", timestamp: int64($0["timestamp"]) ?? 0)
                },
                contacts: (object["contacts"] as? [[String: Any]] ?? []).map {
                    SyncContact(
                        name: $0["name"] as? String ?? "",
                        email: $0["email"] as? String ?? "",
                        phone: $0["phone"] as? String ?? ""
                    )
                },
                attachments: (object["attachments"] as? [[String: Any]] ?? []).map {
                    SyncAttachment(
                        hash: $0["hash"] as? String ?? "",
                        name: $0["name"] as? String ?? "",
                        kind: $0["kind"] as? String ?? ""
                    )
                },
                createdAt: int64(object["createdAt"]) ?? 0
            )
        }
    }

    static func reminders(from data: Data) -> [SyncReminderRecord] {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let array = root["reminders"] as? [[String: Any]] else { return [] }
        return array.compactMap { object in
            guard let id = object["id"] as? String, !id.isEmpty else { return nil }
            return SyncReminderRecord(
                id: id,
                updatedAt: int64(object["updatedAt"]) ?? 0,
                isDeleted: object["deleted"] as? Bool ?? false,
                text: object["text"] as? String ?? "",
                details: object["details"] as? String ?? "",
                triggerAt: int64(object["triggerAt"]) ?? 0,
                leadMinutesBefore: int(object["leadMinutesBefore"]),
                repeatAmount: int(object["repeatAmount"]),
                repeatUnit: string(object["repeatUnit"]),
                tag: object["tag"] as? String ?? "",
                tagEmoji: object["tagEmoji"] as? String ?? "",
                isDone: object["done"] as? Bool ?? false,
                createdAt: int64(object["createdAt"]) ?? 0
            )
        }
    }

    // MARK: - Merging

    static func merge(local: [SyncTaskRecord], remote: [SyncTaskRecord]) -> [SyncTaskRecord] {
        var byID: [String: SyncTaskRecord] = [:]
        for record in local + remote {
            byID[record.id] = byID[record.id].map { winner($0, record) } ?? record
        }
        return sorted(Array(byID.values))
    }

    static func merge(local: [SyncReminderRecord], remote: [SyncReminderRecord]) -> [SyncReminderRecord] {
        var byID: [String: SyncReminderRecord] = [:]
        for record in local + remote {
            byID[record.id] = byID[record.id].map { winner($0, record) } ?? record
        }
        return sorted(Array(byID.values))
    }

    /// Newer wins. Then a delete wins over a live row, because a delete is a decision and a stale
    /// edit is not. Then the greater title, then the greater notes, purely so two devices holding
    /// genuinely different rows stamped at the same millisecond still pick the *same* side.
    static func winner(_ a: SyncTaskRecord, _ b: SyncTaskRecord) -> SyncTaskRecord {
        if a.updatedAt != b.updatedAt { return a.updatedAt > b.updatedAt ? a : b }
        if a.isDeleted != b.isDeleted { return a.isDeleted ? a : b }
        if a.title != b.title { return isGreater(a.title, b.title) ? a : b }
        if a.notes != b.notes { return isGreater(a.notes, b.notes) ? a : b }
        return a
    }

    static func winner(_ a: SyncReminderRecord, _ b: SyncReminderRecord) -> SyncReminderRecord {
        if a.updatedAt != b.updatedAt { return a.updatedAt > b.updatedAt ? a : b }
        if a.isDeleted != b.isDeleted { return a.isDeleted ? a : b }
        if a.text != b.text { return isGreater(a.text, b.text) ? a : b }
        if a.details != b.details { return isGreater(a.details, b.details) ? a : b }
        return a
    }

    /// Compares UTF-8 bytes, not strings — the same rule and the same reason as
    /// SyncNoteDocument.isGreater.
    static func isGreater(_ a: String, _ b: String) -> Bool {
        let x = Array(a.utf8)
        let y = Array(b.utf8)
        for index in 0..<min(x.count, y.count) where x[index] != y[index] {
            return x[index] > y[index]
        }
        return x.count > y.count
    }

    /// Ordered by id alone, so the file's bytes don't churn between syncs that changed nothing.
    static func sorted(_ tasks: [SyncTaskRecord]) -> [SyncTaskRecord] { tasks.sorted { $0.id < $1.id } }
    static func sorted(_ reminders: [SyncReminderRecord]) -> [SyncReminderRecord] {
        reminders.sorted { $0.id < $1.id }
    }

    /// What a person should see: tombstones are bookkeeping, not rows.
    static func visible(_ tasks: [SyncTaskRecord]) -> [SyncTaskRecord] { tasks.filter { !$0.isDeleted } }
    static func visible(_ reminders: [SyncReminderRecord]) -> [SyncReminderRecord] {
        reminders.filter { !$0.isDeleted }
    }

    /// Every file the board still points at. What isn't in here is an orphan and can go.
    static func referencedHashes(_ tasks: [SyncTaskRecord]) -> Set<String> {
        Set(tasks.filter { !$0.isDeleted }.flatMap(\.attachments).map(\.hash).filter { !$0.isEmpty })
    }

    // MARK: - JSON plumbing

    private static func json(for task: SyncTaskRecord) -> [String: Any] {
        [
            "id": task.id,
            "updatedAt": task.updatedAt,
            "deleted": task.isDeleted,
            "title": task.title,
            "notes": task.notes,
            "notesRtl": task.notesRtl,
            "priority": task.priority,
            "dueAt": task.dueAt as Any? ?? NSNull(),
            "orderIndex": task.orderIndex,
            "done": task.isDone,
            "archived": task.isArchived,
            "progress": task.progress,
            "completedAt": task.completedAt as Any? ?? NSNull(),
            "blockedBy": task.blockedBySyncID as Any? ?? NSNull(),
            "waitingOnName": task.waitingOnName,
            "waitingOnSince": task.waitingOnSince as Any? ?? NSNull(),
            "waitingOnFollowUpDays": task.waitingOnFollowUpDays as Any? ?? NSNull(),
            "reminderMinutesBefore": task.reminderMinutesBefore as Any? ?? NSNull(),
            "repeatIntervalDays": task.repeatIntervalDays as Any? ?? NSNull(),
            "tags": task.tags,
            "links": task.links.map { ["url": $0.url, "label": $0.label] },
            "actionLog": task.actionLog.map { ["text": $0.text, "timestamp": $0.timestamp] },
            "contacts": task.contacts.map { ["name": $0.name, "email": $0.email, "phone": $0.phone] },
            "attachments": task.attachments.map { ["hash": $0.hash, "name": $0.name, "kind": $0.kind] },
            "createdAt": task.createdAt,
        ]
    }

    private static func json(for reminder: SyncReminderRecord) -> [String: Any] {
        [
            "id": reminder.id,
            "updatedAt": reminder.updatedAt,
            "deleted": reminder.isDeleted,
            "text": reminder.text,
            "details": reminder.details,
            "triggerAt": reminder.triggerAt,
            "leadMinutesBefore": reminder.leadMinutesBefore as Any? ?? NSNull(),
            "repeatAmount": reminder.repeatAmount as Any? ?? NSNull(),
            "repeatUnit": reminder.repeatUnit as Any? ?? NSNull(),
            "tag": reminder.tag,
            "tagEmoji": reminder.tagEmoji,
            "done": reminder.isDone,
            "createdAt": reminder.createdAt,
        ]
    }

    // A JSON null must read back as nothing, not as zero — the writer meant "no due date", and a
    // default would quietly invent one.
    private static func int64(_ value: Any?) -> Int64? { (value as? NSNumber)?.int64Value }
    private static func int(_ value: Any?) -> Int? { (value as? NSNumber)?.intValue }
    private static func string(_ value: Any?) -> String? {
        guard let text = value as? String, !text.isEmpty else { return nil }
        return text
    }
}
